### ISSUE TITLE
Order of empty images in the batch changes the mAP results

### ISSUE THREAD

**Issue by pierrot-lc** [Lightning-AI/torchmetrics#1774]
Hi,
I faced a similar issue of a previous issue (#884), I think there's a corner case not covered yet. When having images with no ground truth, the order of those images in the batch can change the mAP calculation.

## 🐛 Bug

When computing the mAP, if empty images are at the end of the batch, they will be ignored in the computation. But the empty images that are placed before the last non-empty image are taken into account.

### To Reproduce

*empty image first*
```py
import torch
from torchmetrics.detection.mean_ap import MeanAveragePrecision

targets = [
    {
        "boxes": torch.zeros((0, 4), dtype=torch.float32),
        "labels": torch.zeros((0,), dtype=torch.long),
    },
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
    },
]

preds = [
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
        "scores": torch.FloatTensor([0.9, 0.8]),
    },
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
        "scores": torch.FloatTensor([0.9, 0.8]),
    },
]
map = MeanAveragePrecision(box_format="xyxy", iou_type="bbox")
metrics = map(preds, targets)
print(metrics["map_50"])  # 0.5
```

*empty image second*
```py
targets = [
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
    },
    {
        "boxes": torch.zeros((0, 4), dtype=torch.float32),
        "labels": torch.zeros((0,), dtype=torch.long),
    },
]

preds = [
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
        "scores": torch.FloatTensor([0.9, 0.8]),
    },
    {
        "boxes": torch.FloatTensor([[0.1, 0.1, 0.2, 0.2], [0.3, 0.3, 0.4, 0.4]]),
        "labels": torch.LongTensor([1, 2]),
        "scores": torch.FloatTensor([0.9, 0.8]),
    },
]
map = MeanAveragePrecision(box_format="xyxy", iou_type="bbox")
metrics = map(preds, targets)
print(metrics["map_50"])  # 1.0
```

### Expected behavior

From my understanding, the second case should also raise a 0.5 map instead of 1.0.

Also note that the behaviour is good if we swap the predictions and the targets (empty predictions but non-empty ground truth), with a map of 0.5050 in both cases.

### Environment

- TorchMetrics version: `0.11.4` (using `pip install torchmetrics[detection]`)
- Python version: `3.11.2`
- PyTorch version: `2.0.1+cu117`
- OS: `Ubuntu 23.04 x86_64`

### Additional notes

I know that it looks like a strange corner case to face. I'm training a detector on crops of images so sometimes (actually often in my case), the crops are empty. I guess that I cannot be the only one facing such edge cases.

Have a good day and thank you for your time.

**Comment by SkafteNicki** [comment]
Hi @pierrot-lc, thanks for raising this issue and sorry for the long reply.

TLDR: this behaviour is not wrong and the official implementation also works like this.

Long answer:

I am trying to revamp our MeanAveragePrecision implementation (PR https://github.com/Lightning-AI/torchmetrics/pull/1832) to use pycocotools directly and in that process check all issues that has to do with the metric, like this issue. What I am seeing is that using the official `pycocotools` implementation I get the same as you, namely that changing the order affects the metric. This was initially also weird to me, but let my try to explain why it happens, by calculating the MAP value by hand. 

I highly recommend going through a tutorial like [this](https://learnopencv.com/mean-average-precision-map-object-detection-model-evaluation-metric/) on how to manually calculate map metric. First step is that we need to calculate the precision-recall curve in the two cases.

We need to do the calculation per class, but since it is we have the same problem if we only had one class, for simplicity lets just say that everything is one class (lets say class 1). To do the calculation, we start by ordering our `preds` based on their confidence. It is essentially already here we run into problems, because both pairs of predictions have the same confidence e.g in both cases we have two correct predictions with the scores 0.8 and 0.9, but also two wrong predictions with the same scores. Thus, we cannot order the predictions just based on the their confidence, the ordering is also dependent on their input order.

We therefore need to split the precision-recall computation based on the two cases:

Case 1 (wrong prediction first, the right prediction, then wrong, then right):
Conf | Matches | Cumulative TP | Cumulative FP | Precision | Recall |
|-------|---------|-----------------|----------------|----------|-------|
0.9 | TP | 0 | 1 | 0/(0+1) = 0 | 0/2=0 |
0.9 | FP | 1 | 1 | 1/(1+1) = 1/2 | 1/2 |
0.8 | TP | 1 | 2 | 1/(1+2) = 1/3 | 1/2 |
0.8  | FP | 2 | 2 | 2/(2+2) = 1/2 | 2/2 = 1 |

Case 2 (right prediction first, then wrong prediction, the right, then wrong):
Conf | Matches | Cumulative TP | Cumulative FP | Precision | Recall |
|-------|---------|-----------------|----------------|----------|-------|
0.9 | TP | 1 | 0 | 1/(1+0) = 1 | 1/2 |
0.9 | FP | 1 | 1 | 1/(1+1) = 1/2 | 1/2 |
0.8 | TP | 2 | 1 | 2/(2+1) = 2/3 | 2/2 = 1 |
0.8 | FP | 2 | 2 | 2/(2+2) = 1/2 | 2/2 = 1 |

As you can already see from the table the precision-recall values are completely different. I drawn the two cases below (sorry for my drawing skills):

![map](https://github.com/Lightning-AI/torchmetrics/assets/24896311/8bdeb970-701f-4580-ba4f-d91c19bed671)

Next step is calculating the average-precision from the precision-recall curve = the area under the curve. This is done through interpolation, because this is a non-smooth function. You can read more about how the interpolation is done [here](https://manalelaidouni.github.io/Evaluating-Object-Detection-Models-Guide-to-Performance-Metrics.html), but in this specific case it corresponds to the green curves in the figure. The area under the green lines are

```txt
case 1: 1/2 * 1 = 1/2
case 2: 1 * 1/2 + 2/3 * 1/2 = 0.8333 (this would be 1, if we had not simplified by making everything one class)
```

So what can we conclude: If the confidence for each sample is not unique, then the order of the input can have an influence on the final map value. This is probably counter intuitive (also for me) but it is the official way of computing this metric.

The bigger question for me: should torchmetrics fix this? We could deal with this specific case, but on the other hand we want our implementation to be similar to the official implementation so I am not sure we should actually deal with this corner case.

Hope it all makes sense.

**Comment by pierrot-lc** [comment]
Hi @SkafteNicki , thank you very much for your precise answer. It is much clearer now, and I want to assure you that your drawing skills are perfectly fine!

Indeed, the issue doesn't lie in your implementation but rather in the definition of mAP overall. One possible way to avoid this problem is by grouping the samples with the same confidence score when calculating precision and recall. However, as you mentioned, I don't believe this should be fixed since the goal is to adhere to the official implementation.

Once again, thank you for taking the time to explain this to me. It has helped me gain a better understanding of mAP. Have a great day!

**Comment by SkafteNicki** [comment]
Glad to have helped :]
